staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -105,19 +115,52 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 		return record.resp, record.ok, record.err
 	}
 
-	resp, ok, err := a.authenticator.AuthenticateToken(ctx, token)
-	if !a.cacheErrs && err != nil {
-		return resp, ok, err
+	type lookup struct {
+		resp *authenticator.Response
+		ok   bool
 	}
 
-	switch {
-	case ok && a.successTTL > 0:
-		a.cache.set(key, &cacheRecord{resp: resp, ok: ok, err: err}, a.successTTL)
-	case !ok && a.failureTTL > 0:
-		a.cache.set(key, &cacheRecord{resp: resp, ok: ok, err: err}, a.failureTTL)
+	c := a.group.DoChan(key, func() (val interface{}, err error) {
+		// We're leaving the request handling stack so we need to handle crashes
+		// ourselves. Log a stack trace and return a 500 if something panics.
+		defer runtime.HandleCrash(func(interface{}) {
+			val, err = nil, errAuthnCrash
+		})
+		// Detach the context because the lookup may be shared by multiple callers,
```

Comment:
Does this actually work?  If request A starts at time 0 and times out at time 30, would a request B for the same key at time 25 also time out at time 30 (i.e. it times out too soon)?  Do we need to only aggregate the same key over a small period of time?

Reply:
That sounds like it would be working as intended. Isn't this what we want?

Reply:
I guess?  Seems a bit weird from the second caller's perspective to get timed out too soon.  Maybe in practice it does not matter?

Reply:
I like to think of it like this: we are giving this lookup 30 seconds. If you ask for it and we started it 25 seconds ago, this will timeout in 5 seconds, and you'll get 25 seconds back. Everyone's happy.

Reply:
30s seems excessive to me?

Reply:
That's our current webhook default, and half of our default RequestTimeout configured for kube-apiserver. This matches those relatively well. If we want to shorten this, it should be done independent of this PR.